### ISSUE TITLE
Confirm client behaviour if FxA provides sync engine choices for existing users

### ISSUE THREAD

**Issue by data-sync-user** [mozilla/application-services#2776]
We are working on [a proposal](https://docs.google.com/document/d/1L0SY8sKxmRJKg7I-nWJHouhQUvfL5QcdlwglV3bAZ-A/) for users to see the "Choose What to Sync" screen when signing in to sync with an existing account. Our plans so far assume that the Sync client code in Firefox Desktop, Fennec, Fenix and Firefox for iOS will broadly do the right thing in this case, but we need to verify whether that's true.  Specifically, for each sync client we need to answer the question:

* If the user signs in with an existing account, and the FxA login process provides non-empty values for declinedSyncEngines/offeredSyncEngines, will the client correctly persist those sync engine choices into sync storage?

If we don't believe that it will, then we need to investigate how much work would be involved in making it do so. Here's a checklist for the clients we need to investigate, based on what clients we expect to be live in Q2 2020:

* [x] Firefox Desktop Release
* [x] Firefox Desktop ESR
* [x] Fennec
* [x] Fenix
* [x] Firefox for iOS Release (with bespoke FxA client code)
* [x] Firefox for iOS Beta (with shared FxA client code)

--

Summary of explorations of the question "If the user signs in with an existing account, and the FxA login process provides non-empty values for declinedSyncEngines/offeredSyncEngines, will the client correctly persist those sync engine choices into sync storage?":

* All existing clients seem to do the right thing if the user's metaglobal is currently missing/empty.
* If there is an existing metaglobal then different clients behave different, and none behave in what I consider the "right" way from the user's perspective:
  * Several clients will honour options that the user *declines* during signin, but will not pref back on options that the user explicitly re-enables during signin.
  * Fennec will completely ignore the user's choices during signin in favour of the existing choices on the server.

Given the above, I don't think we can show CWTS on every single signin, but I think we can and should show it during signin *for users who do not have an existing metaglobal*. 

┆Issue is synchronized with this [Jira Task](https://jira.mozilla.com/browse/SYNC-918)
┆Story Points: 3
┆Sprint: SYNC - end 2020-04-10

**Comment by mhammond** [comment]
IIUC, none of our implementations handle or understand "offered" engines in any way, shape or form.

**Comment by rfk** [comment]
Good point. IIRC offeredEngines is supposed to help with the possibility of the fxa-content-server not having support for one of the engines supported by the browser, but we haven't had that problem in practice so maybe it doesn't matter.

**Comment by mhammond** [comment]
>  help with the possibility of the fxa-content-server not having support for one of the engines supported by the browser

I think the plan is that everyone will learn to persist that in meta/global, so we can reliably detect the first time a browser is used with a new data-type, so we can offer that data-type in a "custom" CWTS interaction.

ie, IIUC, browsers learning how to handle this is a critical part of getting CWTS correct?

**Comment by rfk** [comment]
> ie, IIUC, browsers learning how to handle this is a critical part of getting CWTS correct?

We should work through those details here, and this issue seems like as good a place as any.

To be clear at the outset: I'm not arguing against doing this work, I think we should definitely land some in-browser improvements! I just want to understand what depends on what.

IIUC, the [current metaglobal](https://github.com/mozilla-services/docs/blob/master/source/sync/storageformat5.rst#metaglobal-record) contains the set of engines that the user is currently syncing (in `engines`) as well as the set of engines that the user has explicitly declined to sync (in `declined`). We've discussed adding a separate `accepted` set as well in order to more reliably persist user preferences.

As a first step, I'm imagining the FxA signin flow doing the following:

* Receive the set of `supported` engines from the browser during fxa status handshake
* Fetch the set of `active` and `declined` engines from the sync servers
* Intersect `undecided = supported - (active + declined)` to find out what engines are being offered for the first time in this browser
* If `undecided` is non-empty, show CWTS so that the user can make a decision.

I'm sure there would be some rough edges, but would this be an incremental improvement over the status quo? It might prompt users to re-make a decision they've already made, for example after a password reset. But TBH that a surprise re-prompting seems better than a surprise lack of prompting, and we can tighten it up incrementally.

**Comment by mhammond** [comment]
There's a final step missing, which is *updating* meta/global - it's not clear who you picture doing that. I assumed the client, because the client will already need to "do the right thing" here in other scenarios - specifically:

* when it discovers empty storage for an already logged in account
and
* when it upgrades and the new version has support for a new collection, but users are already logged in.

The client is going to need to have some involvement here.  Because the clients already store the selected engines locally (at least for supported ones, but they can easily store it for non-supported ones too), the "new account" scenario and the "empty storage" scenarios could be made the same - something like:

1) Account doesn't exist? FxA account flow, passes accepted/declined to the client. Client stores these prefs locally as if it was syncing in the past. The "native" CWTS UI is correct now, even before the first sync.
2) Client finds no storage? It uses its prefs to populate meta/global.
3) Client has a local engine that's not in either "accepted" nor "declined"? Needs to take some action to get consent - either a new FxA entry-point, or using native UI.

All states can go through this entire flow. In the normal case, you'll never take the affirmative steps for (1) *and* (3), but that's just because FxA and the client have ensured the engine lists are the same - but if somehow you did, everything works out (eg, if FxA somehow declined to offer an engine that does exist locally, you'd get *two* CWTS rounds, but you'd get to the right place in the end)

Does that make sense?

**Comment by mhammond** [comment]
(And to be clear, I'm not necessarily advocating for that - we might have better UX if *both* the client and server update m/g at different times, and that would be OK - the above is really just documenting some assumptions I had previously made)

**Comment by rfk** [comment]
> There's a final step missing, which is updating meta/global - it's not clear who you picture doing 
> that. I assumed the client, because the client will already need to "do the right thing" here in other
> scenarios

Right, yes, I also assumed the client would need to do this, if only because we need the client to do those updates in *some* cases already (such as when the user goes in and edits their CWTS preferences in native client UI).

The specific question I had in mind for this bug is: if FxA shows a CWTS screen during the signin process for an existing user, and the user makes choices on that screen that differ from what's currently in metaglobal, will our current clients correctly update metaglobal to reflect the new choices? Or do we need to land client code to handle this case?

If it helps clarify, I am talking about steps 9 and 10 in the flow [proposed in this doc](https://docs.google.com/document/d/1EDzyi4SPf2Ou5Hkt2Ws8kOOIHmcVBCvPA9cNge6UBKk/).

> Client has a local engine that's not in either "accepted" nor "declined"?
> Needs to take some action to get consent - either a new FxA entry-point, or using native UI.

This is the "I'm already signed in to sync, and I get a Firefox update with support for syncing a new datatype" scenario, right?

I haven't through too deeply about that case yet TBH, but I think I lean towards "new native UI". A good item to bring up with @ryanfeeley next time we discuss this.

**Comment by mhammond** [comment]
> The specific question I had in mind for this bug is: if FxA shows a CWTS screen during the signin process for an existing user, and the user makes choices on that screen that differ from what's currently in their metaglobal, will our current clients correctly update metaglobal to reflect the new choices?

Right - that's what I was getting at in my initial comment here - no clients will write "accepted" and IIUC, that's a key part of what needs to happen, right?

**Comment by rfk** [comment]
True. But if they start syncing a new collection then they'll write its metadata into the `engines` list, and if the user explicitly opts out of the new collection then they'll write it into `declined`...so maybe the user won't be prompted again about that engine in the majority of cases?

I suspect that I just don't understand the proposal around "accepted" in enough detail...

**Comment by mhammond** [comment]
We chatted a little on slack, but:

>  won't be prompted again about that engine in the majority of cases?

Yes, that's certainly true in the majority of cases. But there are edge-cases where 'accepted' would prevent showing CWTS at all, whereas not having it would mean we do - and worse, we may need to on multiple devices.

As mentioned up-thread:

> We've discussed adding a separate accepted set as well in order to more reliably persist user preferences. 

This pretty much nails it - 'accepted' is about reliability.

**Comment by mhammond** [comment]
Just one additional complication - the user will be asked about "history", which includes the "forms" and "history" collections. Ditto for "addons", which will include the "addons" collection and a future "storage.sync" (or whatever) collection.

**Comment by rfk** [comment]
> the user will be asked about "history", which includes the "forms" and "history" collections.
> Ditto for "addons", which will include the "addons" collection and a future "storage.sync" (or
> whatever) collection.

Hrm. Maybe as part of the CWTS work we can consolidate some terminology to make this clearer. Is this a 1:N mapping from "engines" to "syncable stores" or is there another layer of indirection involved in our mental model of the system?

**Comment by mhammond** [comment]
I don't even know what an "engine" or a "syncable store" is :( As we started thinking about syncing "containers" it because even more complicated (eg, one engine/store syncing multiple collections is something that doesn't really happen now, but might in that world)

I think it's really difficult to create a unified model - it literally is closer to "In a effort to not overwhelm the user, what things can we logically group for the purposes of the UI?" From the user's POV, this would probably be an "engine" (ie, forms+history is one engine, addons+addon storage is one engine), but this is totally different to what we think of as an engine - both in the desktop and rust worlds.

Worse, I could imagine a future where this changes - "addresses" might not be clear to the user, and "addresses" and "credit-cards" being split doesn't seem that useful - I could imagine a future where, say, a new, say "forms data" category is created that covers both of these. At that point, "form history" would obviously beg to be moved from "history" too?

**Comment by rfk** [comment]
To test current client behaviour here, I made a very hacky branch of FxA that shows CWTS on signin:

  https://github.com/mozilla/fxa/compare/cwts-on-signin-for-science?expand=1

Running against a local server stack with this change, I created an account, then tried signing in on various clients. Observations so far:

* Signing in on Desktop release channel and disabling "prefs" seemed to do the right thing. On its first sync Desktop correctly moved "prefs" into the list of declined engines and wiped my prefs collection.
* Signing in on Fenix and disabling "bookmarks" seemed to do the right thing. On its first sync Fenix correctly moved "bookmarks" into the list of declined engines and wiped my bookmarks collection.

I'll add others here as I test them.

**Comment by rfk** [comment]
I installed and signed in to ESR, turning off "history" on the CWTS screen while leaving "bookmarks" enabled. This appeared to merge the declined list with the one on the server, resulting in my final declined list being ["history", "forms", "bookmarks", "prefs"].

If we choose to show a CWTS screen on signin to older clients, then they will honour any engines tha that the user opts *out* of, but they will not honour any engines that the user opts back *in* to on this screen. That's bad UX, but at least it fails closed (you end up not syncing data that you think should be syncing, rather than syncing that that you think should not be).

On balance we're probably better off not showing the CWTS screen to these older clients in that case.

**Comment by rfk** [comment]
I tried clearning my server-side sync data and signing in again on ESR, again turning off only the "history" collection. It did the right thing in this case, by opting me out of syncing "history" and "forms". other clients connected to the account also seemed to accept the new values written by the new signin as being the canonical data choices for the account.

**Comment by rfk** [comment]
I tried signing in to Fennec with "bookmarks" declined on the server, and opted out of "history" during the signin process. Unfortunately Fennec did not do what I expected here, it discarded the CWTS data from the signin and used the server's version as canonical.

If I try signing in to Fennec with an *empty* server state, then it does correctly persist the data choices I make during the signin process.

**Comment by rfk** [comment]
I tried signing in to Firefox for iOS built from latest master (with the new Rust FxA backend) and disabling "history". Similar to desktop, it merged the list of things I declined in that signin with the list of things declined on the server.

**Comment by rfk** [comment]
I tried signin in with Firefox for iOS built from the v24.x branch, which predates the Rust FxA backend. I see some strange behaviour with it through.

If I sign in to an existing account that already has data on the server, then it seems to behave similarly to other browsers above, merging the list of things I decline on CWTS-during-signin with the list of things already declined on the server.

If I sign in to an account that does not have any data on the server, then it initially writes a meta/global record with an empty list of declined engines. If I force it to do a "sync now" then it updates that with the data choices I made when signing in to the device. This appears to happen both for new accounts, and for existing accounts that have had their storage wiped. I think it's an independent bug in CWTS behaviour on Firefox for iOS and I'll see if I can come up with some STR that are worth filing upstream.
